File: SecretSharing.Lib/SecretSharing.Lib/SecretSharing.Core/QualifiedSubset.cpp
#pragma once
#include "stdafx.h"
#include "QualifiedSubset.h"
using namespace System;
using namespace System::Collections::Generic;
using namespace std;
using namespace SecretSharingCore::Common;
using namespace SecretSharingCore::Algorithms::GeneralizedAccessStructure;
using namespace System::Linq;

QualifiedSubset::QualifiedSubset(){
	this->Parties = gcnew List<Trustee^>();
}
QualifiedSubset::QualifiedSubset(String^ subset){
	this->Parties = gcnew List<Trustee^>();
	try{
		array<String^>^ parties = subset->Split('^');
		for each (String^ p in parties)
		{
			Trustee^ partyObj = gcnew Trustee(p);
			this->Parties->Add(partyObj);
		}
	}
	catch (Exception^ ex) {
		throw gcnew System::Exception("Invalid Qualified subset example of valid subset: 1^2^3");
	}
}
QualifiedSubset::QualifiedSubset(IEnumerable<Trustee^>^ qualifiedPath)
{
	this->Parties = gcnew List<Trustee^>();
	this->Parties->AddRange(qualifiedPath);
}

bool QualifiedSubset::Equals(Object^ obj) { // no "override" here 
	if (obj == nullptr || GetType() != obj->GetType())
		return false;

	QualifiedSubset^ p = dynamic_cast<QualifiedSubset^>(obj);
	if (this->Parties->Count == p->Parties->Count){
		
		for each (Trustee^ var in p->Parties)
		{
			if (!this->Parties->Contains(var))
				return false;
		}
		return true;
	}
	return false;
}
int QualifiedSubset::GetHashCode() { // no "override" here
	int re = 1;
	for each (Trustee^ var in this->Parties)
	{
		re ^= var->GetPartyId();
	}
	return re;
}


String^  QualifiedSubset::ToString(){
	IEnumerable<String^>^ stringified = Enumerable::Select(this->Parties, gcnew Func<Trustee^, String^>(&Trustee::ToString));
	Func<String^, String^, String^>^ func = gcnew Func < String^, String^, String^ >(this, &QualifiedSubset::aggreagate);
	if (Enumerable::Count(stringified) > 0){
		return Enumerable::Aggregate(stringified, func);
	}
	else{
		return "";
	}
}

String^ QualifiedSubset::aggreagate(String^ current, String^ next){
	return current + "," + next;
}

File: SecretSharing.Lib/SecretSharing.OptimalThreshold/Models/QualifiedSubset.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SecretSharing.OptimalThreshold.Models
{

    public class QualifiedSubset :IComparable,ISubset
    {
        public List<Trustee> Parties;
        public QualifiedSubset()
        {
            this.Parties = new List<Trustee>();
        }
        public QualifiedSubset(String subset)
        {
            this.Parties = new List<Trustee>();
            try
            {
                string[] parties = subset.Split('^');
                foreach (var p in parties)
                {
                    Trustee partyObj = new Trustee(p);
                    this.Parties.Add(partyObj);
                }
            }
            catch
            {
                throw new Exception("Invalid Qualified subset example of valid subset: 1^2^3");
            }
        }
       public  QualifiedSubset(IEnumerable<Trustee> qualifiedPath)
        {
            this.Parties = new List<Trustee>();
            this.Parties.AddRange(qualifiedPath);
        }

        public override bool Equals(Object obj)
        { // no "override" here 
            if (obj == null || GetType() != obj.GetType())
                return false;

            QualifiedSubset p = (QualifiedSubset)(obj);
            if (this.Parties.Count == p.Parties.Count)
            {

                foreach (Trustee var in p.Parties)
                {
                    if (!this.Parties.Contains(var))
                        return false;
                }
                return true;
            }
            return false;
        }
        public override int GetHashCode()
        { // no "override" here
            int re = 1;
            foreach (Trustee t in this.Parties)
            {
                re ^= t.GetPartyId();
            }
            return re;
        }

        public static QualifiedSubset GetAllInvolvedParties(IEnumerable<QualifiedSubset> elements)
        {
            var sum = new QualifiedSubset();
            foreach (var qs in elements)
            {
                foreach (var party in qs.Parties)
                {
                    if (!sum.Parties.Contains(party))
                    {
                        sum.Parties.Add(party);
                    }
                }
            }

            return sum;
        }
        public static string DumpElementsIntoSingleString(IEnumerable<QualifiedSubset> sets)
        {
            StringBuilder sb = new StringBuilder();
            foreach (var item in sets)
            {
                sb.AppendFormat(" {0} ", item);
            }
            return sb.ToString();
        }

        string strValue = null;
        public override String ToString()
        {
            if (strValue == null)
            {
                IEnumerable<String> stringified = Parties.Select(po => po.ToString());
                if (stringified.Count() > 0)
                {
                    strValue = "(" + stringified.Aggregate((current, next) => current + "∧" + next) + ")";

                }
                else
                {
                    strValue = "";
                }
            }
            return strValue;
        }


        public int getPartiesCount()
        {
            return this.Parties.Count;
        }


        public int getShareBranchesCount()
        {
            return getPartiesCount();
        }


        public int getPartyId(int index)
        {
            return this.Parties[index].GetPartyId();
        }

        public int CompareTo(object obj)
        {
            if (obj == null || obj.GetType() != this.GetType()) return 0;
            return this.ToString().CompareTo(((QualifiedSubset)obj).ToString());
        }
    }
}
